### PR TITLE
Add fixed for QSFP temperature and EEPROM

### PR DESCRIPTION
**Pull request by AdiFogel**
Issues that needed to be fixed
- QSFP temperature - when one side of the cable was disconnected the link will
be changed to down. In this case, if the second side has disconnected also,
we would still see the temperature on the cable, because we did not detected
the link change.
- QSFP EEPROM - In case cables have been changed between the set_emu_param.sh cycle,
we will not read the new EEPROM because it did not notice a change in the link.
 With this change, we will read it in any case that we have the cable.
- mlxcables soon to be deprecated.
Solution:
- On each loop of the set_emu script, if the port detected a connection, the
   temperature and EEPROM data will be updated
- Using mlxlink command instead of mlxcables
Tests:
- 6 reboots on each cable state (plugged bot side, plugged one side and both
 sides), verify that temperature and EEPROM are set correctly.
- Test case scenarios: while Linux is booting unplug the cable and plug it back.
- Right before reboot, unplug the cable, issue reboot, checked that is marked as
 unplugged, and plug back.

### CODE REVIEW

**Comment by AdiFogel** [comment]
Hi @odrang and @asmaamellanox,

Please review

File: lanserv/mellanox-bf/set_emu_param.sh [hunk 1]
```diff
@@ -443,9 +443,7 @@ if [ ! -s $EMU_PARAM_DIR/eth_bdfs.txt ] && [ ! -s $EMU_PARAM_DIR/ib_bdfs.txt ];
 else
 	bdf_eth=$(head -n 1 $EMU_PARAM_DIR/eth_bdfs.txt)
 	bdf_ib=$(head -n 1 $EMU_PARAM_DIR/ib_bdfs.txt)
-	p0_changed=0
-	p1_changed=0
-
+	mst cable add > /dev/null 2>&1
```
Comment (odrang) [inline lanserv/mellanox-bf/set_emu_param.sh:446]: add comment what the "mst cable add" perform

File: lanserv/mellanox-bf/set_emu_param.sh [hunk 2]
```diff
@@ -566,19 +562,14 @@ get_fw_info() {
 # Only try to detect the cables via "mst cable add" if the link status
 # has changed.
```
Comment (odrang) [inline lanserv/mellanox-bf/set_emu_param.sh:563]: wrong comment, please fix

File: lanserv/mellanox-bf/set_emu_param.sh (hunk 1 above)
Comment (AdiFogel) [reply lanserv/mellanox-bf/set_emu_param.sh:447]: done

File: lanserv/mellanox-bf/set_emu_param.sh (hunk 2 above)
Comment (AdiFogel) [reply lanserv/mellanox-bf/set_emu_param.sh]: done

**Review by asmaamellanox — CHANGES_REQUESTED** [review]
This is not going to work. I used to have "mst cable add" run every time the script runs and we had to change that behavior:
https://github.com/Mellanox/mlx-OpenIPMI/pull/17/commits/c0950782267098467a6d5e1e726aecbc088b8cfc

Please talk to the mft team to find a better solution

**Comment by AdiFogel** [comment]
Hi @asmaamellanox  and @odrang,
The MR was updated and tested according to commit message.

Please review

**Comment by odrang** [comment]
Hi Adi:
on the pull request:

3. mlxcables soon to be deprecated.
Solution:

What is the solution?

File: lanserv/mellanox-bf/set_emu_param.sh [hunk 3]
```diff
@@ -160,26 +160,22 @@ get_qsfp_eeprom_data() {
...
-
-	# strip the raw hex data from byte id
-	sed s/[0-9][0-9][0-9]:// $EMU_PARAM_DIR/temp1 > $EMU_PARAM_DIR/temp2
-
-	# Put all data in one line and rm space between bytes,
-	# then convert it to a binary file.
-	cat $EMU_PARAM_DIR/temp2 | tr -d ' ' | tr -d '\n' | perl -lpe '$_=pack"H*",$_' > $EMU_PARAM_DIR/temp1
+	mlxlink -d $1 --cable --read --page 0 --offset 0 --length 255 \
+	| grep "page\[0\].Byte"  | cut -f 2 -d ":" | tr -d ' ' \
+	| tr -d "\n"  | perl -lpe '$_=pack"H*",$_' >  $EMU_PARAM_DIR/temp1
```
Comment (odrang) [inline lanserv/mellanox-bf/set_emu_param.sh:165]: change the name temp1--> tmp (confusing with the temp which mean temperature)

File: lanserv/mellanox-bf/set_emu_param.sh [hunk 4]
```diff
@@ -426,9 +423,42 @@ fi
 # for the NIC pci. The NIC card can be associated with device
 # class 0200 if it is connected via pcie or class 0207 if it
 # is connected via infiniband.
+# If a port is not connected or if its temperature is reported as
+# "N/A" by FW, then the ipmitool command will display "no reading".
+
 lspci -n | grep 0200 | cut -f 1 -d " " > $EMU_PARAM_DIR/eth_bdfs.txt
 lspci -n | grep 0207 | cut -f 1 -d " " > $EMU_PARAM_DIR/ib_bdfs.txt
 
+update_cables_info()
```
Comment (odrang) [inline lanserv/mellanox-bf/set_emu_param.sh:432]: what is $1? add comment

File: lanserv/mellanox-bf/set_emu_param.sh [hunk 5]
```diff
@@ -443,57 +473,19 @@ if [ ! -s $EMU_PARAM_DIR/eth_bdfs.txt ] && [ ! -s $EMU_PARAM_DIR/ib_bdfs.txt ];
...
-			fi
-		done <$EMU_PARAM_DIR/eth_bdfs.txt
-	fi
-
-	if [ -s $EMU_PARAM_DIR/ib_bdfs.txt ]; then
-		while read bdf; do
-			func=$(echo $bdf | cut -f 1 -d " " | cut -f 2 -d ".")
-			link_status=$(cat /sys/class/net/ib*$func/operstate)
+	update_cables_info $EMU_PARAM_DIR/eth_bdfs.txt
+	update_cables_info $EMU_PARAM_DIR/eth_bdfs.txt
```
Comment (odrang) [inline lanserv/mellanox-bf/set_emu_param.sh:483]: what about case of IB?
eth_bdfs.txt duplicated

File: lanserv/mellanox-bf/set_emu_param.sh [hunk 6]
```diff
@@ -426,9 +423,42 @@ fi
...
+# "N/A" by FW, then the ipmitool command will display "no reading".
+
 lspci -n | grep 0200 | cut -f 1 -d " " > $EMU_PARAM_DIR/eth_bdfs.txt
 lspci -n | grep 0207 | cut -f 1 -d " " > $EMU_PARAM_DIR/ib_bdfs.txt
 
+update_cables_info()
+{
+	if [ -s $1 ]; then
+		while read bdf; do
+			func=$(echo $bdf | cut -f 1 -d " " | cut -f 2 -d ".")
```
Comment (odrang) [inline lanserv/mellanox-bf/set_emu_param.sh:436]: add comments

**Comment by AdiFogel** [comment]
> Hi Adi: on the pull request:
> 
> 3. mlxcables soon to be deprecated.
>    Solution:
> 
> What is the solution?

From the commit message:
Solution:
1. On each loop of the set_emu script, if the port detected a connection, the
   temperature and EEPROM data will be updated
2. Using mlxlink command instead of mlxcables

File: lanserv/mellanox-bf/set_emu_param.sh [hunk 7]
```diff
@@ -426,9 +423,42 @@ fi
...
+			mlxlink -d $bdf --cable --ddm > /dev/null 2>&1
+			# The port is connected
+			if [ $? -eq 0 ]; then	
+				### Get the temperature for QSFP the port ###
+				get_qsfp_temp $bdf "p${func}_temp"
+				### Update the qsfp eeprom fru
+				get_qsfp_eeprom_data $bdf "qsfp${func}_eeprom"
+			# The port is disconnect
+			else
+				echo 2 > $EMU_PARAM_DIR/p$func"_link"
```
Comment (odrang) [inline lanserv/mellanox-bf/set_emu_param.sh:453]: do we need this or it is already handled in line 442?

**Review by odrang — CHANGES_REQUESTED** [review]

**Comment by odrang** [comment]
ok so in that case the alignment of the commit message is not clear ,should be:

3. mlxcables soon to be deprecated.
   Solution:
          - On each loop of the set_emu script, if the port detected a connection, the
          - temperature and EEPROM data will be updated
          - Using mlxlink command instead of mlxcables
4. Tests:
          - reboots on each cable state (plugged bot side, plugged one side and both
            sides), verify that temperature and EEPROM are set correctly.
          - Test case scenarios: while Linux is booting unplug the cable and plug it back.
            Right before reboot, unplug the cable, issue reboot, checked that is marked as
            unplugged, and plug back.

File: lanserv/mellanox-bf/set_emu_param.sh (hunk 4 above)
Comment (AdiFogel) [reply lanserv/mellanox-bf/set_emu_param.sh:433]: added

File: lanserv/mellanox-bf/set_emu_param.sh (hunk 5 above)
Comment (AdiFogel) [reply lanserv/mellanox-bf/set_emu_param.sh]: fixed

File: lanserv/mellanox-bf/set_emu_param.sh (hunk 6 above)
Comment (AdiFogel) [reply lanserv/mellanox-bf/set_emu_param.sh:438]: added

**Review by odrang — APPROVED** [review]
looks good

**Comment by asmaamellanox** [comment]
Please view my comments from https://github.com/Mellanox/mlx-OpenIPMI/commit/1ee492f25f12efacb77dc4cd2916736556ee6cbb

regarding this change.

**Review by asmaamellanox — CHANGES_REQUESTED** [review]
please see my comments in:
https://github.com/Mellanox/mlx-OpenIPMI/commit/1ee492f25f12efacb77dc4cd2916736556ee6cbb

File: lanserv/mellanox-bf/set_emu_param.sh [hunk 8]
```diff
@@ -160,32 +160,58 @@ get_qsfp_eeprom_data() {
...
+	temp=$(mlxlink -d $1 -m 2> /dev/null | grep Temperature | cut -f 2 -d ":" | tr -d " ")
 	if [ "$temp" != "N/A" ]; then
+		temp=$(echo "$temp" | awk -F'[^0-9]+' '{print $1}')
 		echo $temp > $EMU_PARAM_DIR/$2
 	else
 		remove_sensor "$2"
 	fi
 }
 
+# $1 is $EMU_PARAM_DIR/eth_bdfs.txt or $EMU_PARAM_DIR/eth_bdfs.txt
```
Comment (asmaamellanox) [inline lanserv/mellanox-bf/set_emu_param.sh:185]: not a descriptive comment. Please replace this by :
$1 is the full path of the ethernet or infiniband bdfs file.

Comment (AdiFogel) [reply lanserv/mellanox-bf/set_emu_param.sh]: done

**Review by asmaamellanox — APPROVED** [review]

**Pull request merged**